design/seed_vm.md

```diff
@@ -0,0 +1,190 @@
+
+Values of all types are the first-class values. They can be stored in global
+variables, local variables, object properties, list and table fields. They can
+be passed as arguments to functions, and return from functions as well.
+
+## Instruction Basics
+
+All SeedLang instructions are 32 bits fix-sized unsigned integer. There are
+three instruction types (iABC, iABx, iAsBx) and 24 (number from 0 to 23)
+opcodes. The layout of the three instruction types are illustrated as follows:
+
+```text
+         0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+         0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+        -----------------------------------------------------------------
+iABC    | Opcode: 6 |     A: 8      |      B: 9       |      C: 9       |
+        -----------------------------------------------------------------
+iABx    | Opcode: 6 |     A: 8      |              Bx: 18               |
+        -----------------------------------------------------------------
+iAsBx   | Opcode: 6 |     A: 8      |             sBx: 18               |
+        -----------------------------------------------------------------
+```
+
+There are at most 64 opcodes for SeedLang byte code, because the opcode occupies
+the lower 6 bits of instruction.
+
+Instruction fields are encoded as simple unsigned integer values, except for
+sBx. Field sBx represents negative numbers, but it doesn't use 2s complement.
+Instead, it has a bias equal to half the maximum integer that can be represented
+by its unsigned counterpart, Bx. For a field size of 18 bits, Bx can hold a
+maximum unsigned integer value of 262143, and so the bias is 131071 (calculated
+as 262143 >> 1). A signed value of x will be encoded as (x + 131071).
+
+## Instruction Notation
+
+| Notation         | Description                                      |
+| ---------------- | ------------------------------------------------ |
+| PC               | Program Counter                                  |
+| R(A), R(B), R(C) | Register indexed by A, B or C                    |
+| RK(B), RK(C)     | Register or constant indexed by B or C           |
+| Kst(n)           | Constant in the constant list indexed by n       |
+| Gbl[sym]         | Global variable named with the string symbol sym |
+| sBx              | Signed offset for JMP instruction                |
+
+For RK(B) and RK(C), the MSB (most significant bit) of B or C is used to
+indicate if B or C is the index of the register or constant:
+
+- MSB == 0: B or C is the index of the register
+- MSB == 1: (B - 256) or (C - 256) is the index of the constant in the constant
+  list
+
+## Instruction Set
+
+All the SeedLang instructions are listed as follows:
+
+| Opcode | Name     | Description                                         |
+| :----: | -------- | --------------------------------------------------- |
+|   0    | MOVE     | Copy a value between registers                      |
```

Comment:
As I mentioned above, it will be way better to add the code highlighting notation to the instruction names inside this table.